packages/flutter/lib/src/widgets/autofill.dart

```diff
@@ -0,0 +1,182 @@
+///            setState(() { isSameAddress = newValue; });
+///          },
+///        ),
+///        // Again the address fields are grouped together for the same reason.
+///        if (!isSameAddress) AutofillGroup(
+///          child: Column(
+///            children: <Widget>[
+///              TextField(
+///                controller: billingAddress1,
+///                autofillHints: <String>[AutofillHints.streetAddressLine1],
+///              ),
+///              TextField(
+///                controller: billingAddress2,
+///                autofillHints: <String>[AutofillHints.streetAddressLine2],
+///              ),
+///            ],
+///          ),
+///        ),
+///        const Text('Credit Card Information'),
+///        // The credit card number and the security code are grouped together as
+///        // some platforms are capable of autofilling both fields.
+///        AutofillGroup(
+///          child: Column(
+///            children: <Widget>[
+///              TextField(
+///                controller: creditCardNumber,
+///                autofillHints: <String>[AutofillHints.creditCardNumber],
+///              ),
+///              TextField(
+///                controller: creditCardSecurityCode,
+///                autofillHints: <String>[AutofillHints.creditCardSecurityCode],
+///              ),
+///            ],
+///          ),
+///        ),
+///        const Text('Contact Phone Number'),
+///        // The phone number field can still be autofilled despite lacking an
+///        // `AutofillScope`.
+///        TextField(
+///          controller: phoneNumber,
+///          autofillHints: <String>[AutofillHints.telephoneNumber],
+///        ),
+///      ],
+///    );
+///  }
+/// ```
+/// {@end-tool}
+class AutofillGroup extends StatefulWidget {
+  /// Creates a scope for autofillable input fields.
+  ///
+  /// The [child] argument must not be null.
+  const AutofillGroup({
+    Key key,
+    @required this.child,
+  }) : assert(child != null),
+       super(key: key);
+
+  /// Returns the closest [AutofillScope] which encloses the given context.
+  ///
+  /// In order to interact with the platform's autofill mechanism,
+  /// [AutofillTrigger]s need to call [AutofillScope.attach] on their closest
+  /// [AutofillScope], instead of calling [TextInputClient.attach].
+  static _AutofillGroupState of(BuildContext context) {
+    final _AutofillScope scope = context.dependOnInheritedWidgetOfExactType<_AutofillScope>();
+    return scope?._scope;
+  }
+
+  /// {@macro flutter.widgets.child}
+  final Widget child;
+
+  @override
+  _AutofillGroupState createState() => _AutofillGroupState();
+}
+
+class _AutofillGroupState extends State<AutofillGroup> with AutofillScopeMixin {
+  @override
+  Iterable<AutofillClient> get autofillClients {
+    final List<AutofillClient> clients = <AutofillClient>[];
+    void visit(Element element) {
+      if (element is AutofillScope)
```

Comment:
The `is` type checks in this method are odd. Can the AutofillClients just register with the Group instead of walking the tree to find them?

Reply:
Also, what happens when a new AutofillClient is added as a child? How does that re-trigger the walk?

Reply:
Register when the widget is built? 
- If the widgets are built in a weird order or one of them rebuilds the platform may get confused (we may be able to work this around by properly grouping autofillable input fields).
- The implementation of `Build` should try to be as sideeffect-free as possible.

Reply:
I thought we solved the ordering problem by grouping related text input widgets and then in that group ordering didn't matter anymore?

I would have expected them to be stateful and register during initState and unregister in dispose.

Reply:
If a client gets reparented to a different `AutofillScope` then it won't have a chance to notify its old scope if we do that in `initState` and `dispose` I think? There's no `activate` method on a `State` object so I think we'll have to do that in `build`.

As for the ordering, with proper grouping the order doesn't seem to matter to iOS password autofill, but on Android the autofill service should still be able to see the order so I can't really say for sure that order doesn't matter if a custom autofill service is used.

Reply:
You could establish a dependency to the Autofill scope and then your `didChangeDependencies` would get called.